File: 621.task-scheduler.go
/*
 * @lc app=leetcode id=621 lang=golang
 *
 * [621] Task Scheduler
 */

// @lc code=start
// Greedy strategy: run the type of tasks with max. count first, separated by other types of tasks
// T: O(n)
// S: O(s), s := the number of task types
func leastInterval(tasks []byte, n int) int {
    var cnts [26]int
	maxCnt := 0
	numOfTasksWithMaxCnt := 0
	for _, task := range tasks {
		ci := task - 'A'
		cnts[ci]++
		if cnts[ci] == maxCnt {
			numOfTasksWithMaxCnt++
		} else if cnts[ci] > maxCnt {
			maxCnt = cnts[ci]
			numOfTasksWithMaxCnt = 1
		}
	}

	// Example: if A has the max. count and n = 2, then we should schedule like [A, x, x, A, x, x, A, ...],
	// where x is any other types of tasks. Consider the subarray [x, x] which has length n. If there exist 
	// other types of tasks also having the max. count, some of the subarray slots must be used by these 
	// most frequent tasks. Using the number of tasks with smaller counts and the number of slots for 
	// these less frequent tasks to get the answer, we can get the number of idle slots, and then the answer.
	numOfSlotsForLessFreqTasks := (maxCnt - 1) * (n - (numOfTasksWithMaxCnt - 1))  // minus 1 since one of the most frequent tasks like A is used
	numOfLessFreqTasks := len(tasks) - maxCnt * numOfTasksWithMaxCnt
	numOfIdleSlots := numOfSlotsForLessFreqTasks - numOfLessFreqTasks
	if numOfIdleSlots < 0 {
		// IMPORTANT: If numOfIdleSlots becomes negative here, it means there are more 
		// less-frequent tasks to separate the same type of most-frequent tasks, so 
		// there won't be any idle slots
		// numOfIdleSlots = 0
		return len(tasks)
	}
	return len(tasks) + numOfIdleSlots
}
// @lc code=end

File: 621.task-scheduler.py
#
# @lc app=leetcode id=621 lang=python3
#
# [621] Task Scheduler
#
from typing import List

# @lc code=start
class Solution:
    # T: O(len(tasks))
    # S: O(n_task_types)
    def leastInterval(self, tasks: List[str], n: int) -> int:
        freqs = [0] * 26
        OFFSET = ord('A')
        n_tasks = len(tasks)
        max_freq = 0
        n_tasks_with_max_freq = 0

        for t in tasks:
            ci = ord(t) - OFFSET
            freqs[ci] += 1
            if freqs[ci] > max_freq:
                max_freq = freqs[ci]
                n_tasks_with_max_freq = 1
            elif freqs[ci] == max_freq:
                n_tasks_with_max_freq += 1

        n_slots_for_less_freq_tasks = (max_freq - 1) * (n - (n_tasks_with_max_freq - 1))
        n_tasks_with_less_freqs = n_tasks - max_freq * n_tasks_with_max_freq
        n_slots_idle = max(0, n_slots_for_less_freq_tasks - n_tasks_with_less_freqs)
        
        return n_tasks + n_slots_idle

# @lc code=end

